src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp

```diff
@@ -241,4 +276,227 @@ bool RenderFillExtrusionLayer::queryIntersectsFeature(const GeometryCoordinates&
+            /* .base_t = */ std::get<0>(binders.get<FillExtrusionBase>()->interpolationFactor(zoom)),
+            /* .height_t = */ std::get<0>(binders.get<FillExtrusionHeight>()->interpolationFactor(zoom)),
+            /* .color_t = */ std::get<0>(binders.get<FillExtrusionColor>()->interpolationFactor(zoom)),
+            /* .pattern_from_t = */ std::get<0>(binders.get<FillExtrusionPattern>()->interpolationFactor(zoom)),
+            /* .pattern_to_t = */ std::get<0>(binders.get<FillExtrusionPattern>()->interpolationFactor(zoom)),
+            /* .pad = */ 0,
+            0,
+            0};
+
+        const FillExtrusionDrawableTilePropsUBO tilePropsUBO = {
+            /* pattern_from = */ patternPosA ? util::cast<float>(patternPosA->tlbr()) : std::array<float, 4>{0},
+            /* pattern_to = */ patternPosB ? util::cast<float>(patternPosB->tlbr()) : std::array<float, 4>{0},
+        };
+
+        // If we already have drawables for this tile, update them.
+        if (tileLayerGroup->getDrawableCount(drawPass, tileID) > 0) {
+            // Just update the drawables we already created
+            tileLayerGroup->observeDrawables(drawPass, tileID, [&](gfx::Drawable& drawable) {
+                auto& uniforms = drawable.mutableUniformBuffers();
+                uniforms.createOrUpdate(
+                    FillExtrusionLayerTweaker::FillExtrusionTilePropsUBOName, &tilePropsUBO, context);
+                uniforms.createOrUpdate(
+                    FillExtrusionLayerTweaker::FillExtrusionInterpolateUBOName, &interpUBO, context);
+            });
+            continue;
+        }
+
+        const auto uniformProps = vertexAttrs.readDataDrivenPaintProperties<FillExtrusionBase,
+                                                                            FillExtrusionColor,
+                                                                            FillExtrusionHeight,
+                                                                            FillExtrusionPattern>(binders, evaluated);
+
+        const auto shader = std::static_pointer_cast<gfx::ShaderProgramBase>(
+            shaderGroup->getOrCreateShader(context, uniformProps));
+        if (!shader) {
+            assert(false);
+            continue;
+        }
+
+        // The non-pattern path in `render()` only uses two-pass rendering if there's translucency.
+        // The pattern path always uses two passes.
+        const auto doDepthPass = (!opaque || hasPattern);
+
+        if (doDepthPass && !depthBuilder) {
+            if (auto builder = context.createDrawableBuilder(layerPrefix + "depth")) {
+                builder->setShader(shader);
+                builder->setIs3D(true);
+                builder->setEnableStencil(false);
+                builder->setEnableColor(false);
+                builder->setRenderPass(drawPass);
+                builder->setCullFaceMode(gfx::CullFaceMode::backCCW());
+                depthBuilder = std::move(builder);
```

Comment:
I see the reason why we have two builders but the extra copies we're doing here just to have different state isn't great. Can we have a single builder and get shared vertex buffers working? ie Two or more drawables sharing the same VAO but configured with different state.

Reply:
If we are going to do that, can we do it at a separate task?  Be happy to schedule it very soon.

Reply:
I'll make an issue for it.

Reply:
Tracking this in #1300, OK to leave this for now.